### PR TITLE
Chore: Refacto

### PR DESCRIPTION
**Pull request by tekkac**
- Use VintageComponent with interface
- switch constants and CarbonVintage to models 
- implemented set_project_carbon (wip)
- implemented set_vintage with yearly_absorptions

TODO:
 - fix TO-DOs
 - fix access control
 - fix or remove commented tests
 - implement rebased vs failed in CarbonVintage
 - add missing events

### CODE REVIEW

File: src/components/vintage/vintage.cairo [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+                    break ();
+                }
+                let vintage: CarbonVintage = CarbonVintage {
+                    year: index.into(), supply: 0, failed: 0, status: CarbonVintageType::Projected,
+                };
+                // [Effect] Store values
+                self.Vintage_vintages.write(index.into(), vintage);
+                index += 1;
+            };
+        }
```
Comment (julienbrs) [inline src/components/vintage/vintage.cairo:224]: Do we really need that initializer? If we decide to go on a 2-steps minter, then, we would initialize and set vintages after the end of raise of funds. No need to initialize before the end.

File: src/components/vintage/vintage.cairo [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+            let mut project_carbon: u128 = 0;
+
+            // let mut index = self.Vintage_vintages_len.read();
+            // while index > 0 {
+            //     index -= 1;
+            //     let vintage = self.Vintage_vintages.read(index.into());
+            //     if vintage.status != CarbonVintageType::Unset {
+            //         project_carbon += vintage.supply.into();
+            //     }
+            // };
```
Comment (julienbrs) [inline src/components/vintage/vintage.cairo:67]: Why not calculating `Vintage_project_carbon` as you did here, and remove the storage var? However we should change that storage value when we update a Vintage with `rebase_vintage` or `rebase_vintage`

**Review by julienbrs — COMMENTED** [review]
Is `Vintage_project_carbon` the total amount of carbon for the project, or only the supply for the vintage?  It seems that it is the total amount,

**Comment by tekkac** [comment]
> Is `Vintage_project_carbon` the total amount of carbon for the project, or only the supply for the vintage? It seems that it is the total amount,

Yes, it _should_ be the total amount as a specific vintage supply is gotten from the CarbonVintage data.
Here, I wanted to set the total amount before setting the vintage (set_vintage call) for the updatable minting process.
set_vintage might need to check the sum of vinatge supplies is equal to project_carbon

File: src/components/vintage/vintage.cairo (hunk 2 above)
Comment (tekkac) [reply src/components/vintage/vintage.cairo:67]: indeed, we still need to figure out if we need this storage var.
The issue is to make it work with the whole process: minting, updating the project carbon before minting and rebasing.
This implementation might not be the final one.

**Pull request merged**